src/io/flutter/run/common/CommonTestConfigUtils.java

```diff
@@ -26,45 +32,76 @@
  */
 public abstract class CommonTestConfigUtils {
   /**
-   * Regex that matches customized versions of the Widget test function from package:flutter_test/src/widget_tester.dart.
-   * <p>
-   * This will match all test methods with names that start with 'test', optionally
-   * have additional text in the middle, and end with 'Widgets'.
+   * How the Dart Analysis Server identifies runnable tests in the {@link Outline} it generates.
    */
-  public static final Pattern WIDGET_TEST_REGEX = Pattern.compile("test([A-Z][A-Za-z0-9_$]*)?Widgets");
+  private static final String UNIT_TEST_TEST = "UNIT_TEST_TEST";
 
-  public abstract TestType asTestCall(@NotNull PsiElement element);
+  /**
+   * How the Dart Analysis Server identifies runnable test groups in the {@link Outline} it generates.
+   */
+  private static final String UNIT_TEST_GROUP = "UNIT_TEST_GROUP";
 
   public static String convertHttpServiceProtocolToWs(String url) {
     return StringUtil.trimTrailing(
       url.replaceFirst("http:", "ws:"), '/') + "/ws";
   }
 
+  public abstract TestType asTestCall(@NotNull PsiElement element);
+
+  /**
+   * Gets the elements from the outline that are runnable tests.
+   */
+  protected Map<DartCallExpression, TestType> getTestsFromOutline(@NotNull PsiFile file) {
+    final Project project = file.getProject();
+    final FlutterOutline outline = OpenEditorOutlineService.getInstance(project).get(file.getVirtualFile().getPath());
```

Comment:
consider adding a method to the outline editor that goes from a VirtualFile to an outline instead of a String.
Also, do we care about cases where the PsiFile and the Outline are out of sync? For example, you can check that their links are consistent. There was some logic in the outline rendering code that makes sure the lengths are the same or can be made the same by transforming the locations. It could make sense to add that logic to the OpenEditorOutlineService so we are consistent about the cases where the outline is slightly out of date so shouldn't be used. The outlines become consistent a second after an edit but if you are displaying visualizations of where tests occur in the source editors in the future then the mismatch will be noticeable.

Reply:
IntelliJ appears to be taking care of consistency here under the hood.  In the current state of this PR, behavior is the same as we get at head.  

Reading the [details](https://www.jetbrains.org/intellij/sdk/docs/tutorials/custom_language_support/line_marker_provider.html#even-more-technical-details) of how the line markers pass works, it looks like it is possible for the pass it takes to take longer than the outline update.

However, there's currently no way to ask IntelliJ to check for runnable tests again after getting the new outline.  The fix for that is to enable the line marker pass to listen to the outline service the same way that the indent highlighting pass does.  Any logic involving outdated outlines will need to go there where we can make another pass instead of here where we can't.